horovod/common/operations.cc

```diff
@@ -1000,32 +1338,118 @@ bool RunLoopOnce(HorovodGlobalState& state, MPIContext& ctx, bool is_coordinator
+          invalid_in_queue = true;
+          uncached_in_queue = true;
+        } else {
+          uncached_in_queue = true;
+        }
+      }
+    }
+  }
+
+  // Flag indicating that the background thread should shut down.
+  bool should_shut_down = state.shut_down;
+
+  if (is_coordinator) {
+    // Check for stalled tensors.
+    if (state.perform_stall_check &&
+        std::chrono::steady_clock::now() - state.last_stall_check >
+            std::chrono::seconds(state.stall_warning_time_seconds)) {
+      should_shut_down |= CheckForStalledTensors(state, ctx);
+      state.last_stall_check = std::chrono::steady_clock::now();
+    }
+  }
+
+  if (state.response_cache.capacity() > 0) {
+    // Obtain common cache hits and cache invalidations across workers. Also,
+    // determine if any worker has uncached messages in queue or requests
+    // a shutdown. This function removes any invalid cache entries, if they
+    // exist.
+    GetCommonCacheAndState(cache_hits, invalid_bits, uncached_in_queue,
+                           invalid_in_queue, should_shut_down, state, ctx);
+
+    {
+      // Get lock in order to safely replace messages to global queue
+      std::lock_guard<std::mutex> guard(state.mutex);
+
+      // Remove uncommon cached tensors from queue and replace to state
+      // queue for next cycle.
+      size_t num_messages = message_queue.size();
+      for (size_t i = 0; i < num_messages; ++i) {
+        auto message = message_queue.front();
+        if (!(state.response_cache.cached(message) ==
+              ResponseCache::CacheState::HIT) ||
+            cache_hits.find(state.response_cache.peek_cache_bit(message)) !=
+                cache_hits.end()) {
+          message_queue.push(std::move(message));
+        } else {
+          state.message_queue.push(std::move(message));
+        }
+        message_queue.pop();
+      }
     }
   }
 
   if (!message_queue.empty()) {
     LOG(DEBUG, state.rank) << "Sent " << message_queue.size() << " messages";
   }
 
-  // Flag indicating that the background thread should shut down.
-  bool should_shut_down = state.shut_down;
+  if (state.response_cache.capacity() > 0) {
+    if (cache_hits.size() > 0 && !uncached_in_queue) {
+      // If only cached messages in queue, use fast coordination path.
+      RunBypass(message_queue, cache_hits, state, ctx);
+      return !should_shut_down;
+    } else if (!uncached_in_queue) {
+      // Quick return if there are no uncached messages in any worker queue.
+      return !should_shut_down;
+    }
+  }
 
   // Collect all tensors that are ready to be reduced. Record them in the
   // tensor count table (rank zero) or send them to rank zero to be
   // recorded (everyone else).
   std::vector<std::string> ready_to_reduce;
   if (is_coordinator) {
     while (!message_queue.empty()) {
+      // Skip cached messages.
+      if (state.response_cache.capacity() > 0 &&
+          state.response_cache.cached(message_queue.front()) ==
```

Comment:
If we're just skipping these, can we skip them on line 1415 and not have this code?

Reply:
Done.